// aes-password/build/main/lib/utils/Deriver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Deriver = void 0;
const crypto_1 = require("crypto");
const HMAC_KEY_SIZE = 32;
const PASSWORD_KEY_SIZE = 32;
class Deriver {
    static pbkdf2({ password, salt, rounds, bits, }) {
        return new Promise((resolve, reject) => {
            (0, crypto_1.pbkdf2)(password, salt, rounds, bits / 8, 'sha256', (err, key) => {
                if (err)
                    return reject(err);
                resolve(key);
            });
        });
    }
    static deriveFromPassword({ password, salt, rounds, }) {
        return new Promise(async (resolve, reject) => {
            const bits = (PASSWORD_KEY_SIZE + HMAC_KEY_SIZE) * 8;
            if (!password)
                reject(new Error('Failed deriving key: Password must be provided'));
            if (!salt)
                reject(new Error('Failed deriving key: Salt must be provided'));
            if (!rounds || rounds <= 0 || typeof rounds !== 'number')
                reject(new Error('Failed deriving key: Rounds must be greater than 0'));
            try {
                const raw = await this.pbkdf2({ password, salt, rounds, bits });
                const derivedKeyHex = raw.toString('hex');
                const dkhLength = derivedKeyHex.length;
                const keyBuffer = Buffer.from(derivedKeyHex.substr(0, dkhLength / 2), 'hex');
                const hmacHex = Buffer.from(derivedKeyHex.substr(dkhLength / 2, dkhLength / 2), 'hex');
                resolve({
                    salt: salt,
                    key: keyBuffer,
                    rounds: rounds,
                    hmac: hmacHex,
                });
            }
            catch (e) {
                reject(e);
            }
        });
    }
}
exports.Deriver = Deriver;
